### PR TITLE
Allow to diasble OpenShift features

### PR DESCRIPTION
**Pull request by mburke5678**
https://trello.com/c/2DisRDjm/773-document-allow-to-disable-openshift-features

### CODE REVIEW

**Comment by mburke5678** [comment]
What happens when a feature is disabled? 
* The feature does not appear in console? 
* The CLI presents an error message?

What happens if an app relies on a feature that gets disabled? (is this a concern?)
The user needs to disable the feature on the master and on each and every node in the cluster one-by-one?
What happens with new nodes? Will they pick up that the feature is disabled in master and default to disable that feature on new nodes?

**Comment by soltysh** [comment]
> What happens when a feature is disabled?
> 
>     The feature does not appear in console?

Assuming you'll disable builds, you'll see the builds tab in the web console. It'll only be empty,
similarly other user facing resources. But it will be dependent on the feature you disable. 

>     The CLI presents an error message?

You'll get a not found error. 

>     What happens if an app relies on a feature that gets disabled? (is this a concern?)

The app might be broken, but that will depend on the app itself. Not a concern. 

>     The user needs to disable the feature on the master and on each and every node in the cluster one-by-one?

Each feature is different. Some are disabled on an API (eg. resources, audit), some are applied on o node level (eg. any kubelet related). Where it's updated depends on just that. 

>     What happens with new nodes? Will they pick up that the feature is disabled in master and default to disable that feature on new nodes?

See the previous answer.

**Comment by mburke5678** [comment]
@soltysh 
>> Each feature is different. Some are disabled on an API (eg. resources, audit), some are applied on o node level (eg. any kubelet related). Where it's updated depends on just that.
Is there a simple way of indicating whether a feature needs to be disabled on the master or the nodes?

File: admin_guide/disabling_features.adoc [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+For example, you could turn off new features for production clusters 
+while leaving the features on for test clusters where you can 
+fully test them. Or, prevent access to alpha, beta, or Technology Preview features. 
+
+If you disable a feature that appears in the web console, you might see that feature, but
+no objects are listed. For example, if you disable builds, you can see the Builds tab in the web console. 
+but there are no builds present. 
+
+If you attempt to use commanfds associated with a disabled feature, such as `oc build`, {product-title}
```
Comment (soltysh) [inline admin_guide/disabling_features.adoc]: s/commanfds/commands

Comment (soltysh) [reply admin_guide/disabling_features.adoc]: There's no `oc build`, `oc start-build` does exists, though.

File: admin_guide/disabling_features.adoc [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+====
+If you disable a feature which any of the applications in the cluster relies on, the application might not
+function properly, depending upon the feature disabled and how the application uses that feature.
+====
+
+Feature gates use a `key=value` pair in the master configuration file
+(*_/etc/origin/master/master-config.yaml_*) and the 
+and kubelet configuration files that describe the feature you want to block.
+
+For exxample, the following code turns off the Huge Pages feature:
```
Comment (soltysh) [inline admin_guide/disabling_features.adoc]: s/exxample/example

**Review by soltysh — COMMENTED** [review]
I left you some comments, but overall lgtm.

**Comment by mburke5678** [comment]
@wsun1 PTAL

**Comment by mburke5678** [comment]
@geliu2016 PTAL

File: admin_guide/disabling_features.adoc [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+. Restart the {product-title} master service to apply the changes.
++
+ifdef::openshift-enterprise[]
+----
+# systemctl restart atomic-openshift-master-api atomic-openshift-master-controllers
+----
+endif::[]
+ifdef::openshift-origin[]
+----
+# systemctl restart origin-master-api origin-master-controllers
```
Comment (geliu2016) [inline admin_guide/disabling_features.adoc]: 'systemctl restart API, controllers' is not be supported in ocp 310.

**Comment by geliu2016** [comment]
@mburke5678 ， opened bug(1576156) after review it, thx

File: admin_guide/disabling_features.adoc (hunk 3 above)
Comment (vikram-redhat) [inline admin_guide/disabling_features.adoc]: Per bug from @geliu2016 'systemctl restart API, controllers' is not be supported in ocp 310.

Suggestions for improvement: 
change to the 'master-restart' command that new design in 310, of course, this command seems not the final solution, anyway, we need to update it follow the design of 310.

**Comment by mburke5678** [comment]
@openshift/team-documentation PTAL

File: admin_guide/disabling_features.adoc [hunk 4]
```diff
@@ -0,0 +1,276 @@
+[[admin-guide-disable-feature]]
+= Disabling Features using Feature Gates
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: s/using/Using

File: admin_guide/disabling_features.adoc [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+:prewrap!:
+
+toc::[]
+
+
+
+As an administrator, you can turn off specific features to specific nodes or to the entire platform
+using _feature gates_.
+
+For example, you could turn off new features for production clusters
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: s/could/can

Comment (ahardin-rh) [reply admin_guide/disabling_features.adoc]: strange line breaks in this paragraph

File: admin_guide/disabling_features.adoc [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+As an administrator, you can turn off specific features to specific nodes or to the entire platform
+using _feature gates_.
+
+For example, you could turn off new features for production clusters 
+while leaving the features on for test clusters where you can 
+fully test them. 
+
+If you disable a feature that appears in the web console, you might see that feature, but
+no objects are listed. For example, if you disable builds, you can see the *Builds* tab in the web console. 
+but there are no builds present.
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: web console, but there

File: admin_guide/disabling_features.adoc [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+endif::[]
+ifdef::openshift-origin[]
+----
+# systemctl restart origin-node
+----
+endif::[]
+
+To re-enable a disabled feature, edit the master and node configuration files to remove the `<feature_name>=false`
+and restart the master and node services.
+
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: Missing section ID for List of Feature Gates

File: admin_guide/disabling_features.adoc [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+| *CPUManager*
+| Enables container-level CPU affinity support. For more information, 
+see xref:../scaling_performance/using_cpu_manager.adoc#scaling-performance-using-cpu-manager[Using CPU Manager].
+
+| *CRIContainerLogRotation*
+| Enables container log rotation for the CRI container runtime.
+
+| *CSIPersistentVolume*
+| Enables discovering and mounting volumes provisioned through a CSI (Container Storage Interface) compatible volume plugin. For more information,
+see the link:https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md[CSI Volume Plugins in Kubernetes Design Doc].
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: CSI Volume Plug-ins in Kubernetes Design Documentation  (I see the actual doc has the title you use,  but we should apply our standards to it anyway)

File: admin_guide/disabling_features.adoc [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+| Enables the scheduler to cache equivalence of nodes when scheduling Pods.
+
+| *ExperimentalCriticalPodAnnotation*
+| Enables annotating specific pods as critical so that their scheduling is guaranteed.
+
+| *ExperimentalHostUserNamespaceDefaultingGate*
+| Enables the disabling of user namespaces. This is for containers that are using other host projects, host mounts, or containers that are privileged or using specific non-project capabilities, such as MKNODE, SYS_MODULE, and so forth. This should only be enabled if user project remapping is enabled in the Docker daemon.
+
+| *GCERegionalPersistentDisk*
+| Enables the xref:../install_config/persistent_storage/persistent_storage_gce.adoc#install-config-persistent-storage-persistent-storage-gce[GCE
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: weird line break here

File: admin_guide/disabling_features.adoc [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+| *HugePages*
+| Enables the allocation and consumption of pre-allocated xref:../scaling_performance/managing_hugepages.adoc#scaling-performance-managing-huge-pages[huge pages].
+
+| *HyperVContainer*
+| Enables Hyper-V isolation for Windows containers.
+
+| *Intializers*
+|Enables the xref:../architecture/additional_concepts/dynamic_admission_controllers.adoc#architecture-additional-concepts-dynamic-admission-controllers[dynamic admission control] 
+as an extension to the built-in admission controllers.
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: another weird line break

File: admin_guide/disabling_features.adoc [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+
+| *Intializers*
+|Enables the xref:../architecture/additional_concepts/dynamic_admission_controllers.adoc#architecture-additional-concepts-dynamic-admission-controllers[dynamic admission control] 
+as an extension to the built-in admission controllers. 
+
+| *LocalStorageCapacityIsolation*
+| Enables the consumption of local ephemeral storage and the `sizeLimit` property of an *emptyDir* volume.
+
+| *MountContainers*
+| Enables using utility containers on host as the volume mount.
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: s/on host/on the host

File: admin_guide/disabling_features.adoc [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+| *PersistentLocalVolumes*
+| Enables the usage of xref:../install_config/configuring_local.adoc#install-config-configuring-loca[local volume] pods. 
+Pod affinity has to be specified if requesting a local volume.
+
+| *PodPriority*
+| Enables the xref:../admin_guide/scheduling/descheduler.adoc#admin-guide-descheduler[descheduling] and preemption of pods based on their priorities.
+
+| *ReadOnlyAPIDataVolumes*
+| Set xref:../dev_guide/secrets.adoc#dev-guide-secrets[secrets], xref:../dev_guide/configmaps.adoc#dev-guide-configmaps[ConfigMap], 
+DownwardAPI and xref:../dev_guide/projected_volumes.adoc#dev-guide-projected-volumes[projected volumes] to be mounted in read-only mode.
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: DownwardAPI, and

File: admin_guide/disabling_features.adoc [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+
+| *PodPriority*
+| Enables the xref:../admin_guide/scheduling/descheduler.adoc#admin-guide-descheduler[descheduling] and preemption of pods based on their priorities.
+
+| *ReadOnlyAPIDataVolumes*
+| Set xref:../dev_guide/secrets.adoc#dev-guide-secrets[secrets], xref:../dev_guide/configmaps.adoc#dev-guide-configmaps[ConfigMap], 
+DownwardAPI and xref:../dev_guide/projected_volumes.adoc#dev-guide-projected-volumes[projected volumes] to be mounted in read-only mode. 
+
+| *ResourceLimitsPriorityFunction*
+| Enables a xref:../admin_guide/scheduling/scheduler.adoc#[admin-guide-scheduler[scheduler] priority function that assigns a lowest possible score of 1 to a node that satisfies at least one of the input pod CPU and memory limits. The intent is to break ties between nodes with same scores.
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: `1`

File: admin_guide/disabling_features.adoc [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+or xref:../architecture/additional_concepts/storage.adoc#persistent-volume-claims[persistent volume claim] objects if they are still being used.
+
+| *StreamingProxyRedirects*
+| Instructs the API server to intercept and follow redirects from the backend kubelet for streaming requests. 
+
+| *SupportIPVSProxyMode*
+| Enables providing in-cluster service load balancing using IP virtual servers.
+
+| *SupportPodPidsLimit*
+| Enables support for limiting PIDs in pods.
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: what are PIDs?

File: admin_guide/disabling_features.adoc [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+| *TaintBasedEvictions*
+| Enables evicting pods from nodes based on xref:../admin_guide/scheduling/taints_tolerations.adoc#admin-guide-taints[taints on nodes and tolerations on pods].
+
+| *TaintNodesByCondition*
+| Enables automatic xref:../admin_guide/scheduling/taints_tolerations.adoc#admin-guide-taints[tainting nodes] based on node conditions.
+
+| *TokenRequest*
+| Enables the TokenRequest endpoint on service account resources.
+
+| *VolumeScheduling**
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: Extra *

File: admin_guide/disabling_features.adoc [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+Persistent Disk] feature.
+
+| *HugePages*
+| Enables the allocation and consumption of pre-allocated xref:../scaling_performance/managing_hugepages.adoc#scaling-performance-managing-huge-pages[huge pages].
+
+| *HyperVContainer*
+| Enables Hyper-V isolation for Windows containers.
+
+| *Intializers*
+|Enables the xref:../architecture/additional_concepts/dynamic_admission_controllers.adoc#architecture-additional-concepts-dynamic-admission-controllers[dynamic admission control]
```
Comment (ahardin-rh) [inline admin_guide/disabling_features.adoc]: In some entries, there is a space between the text and the |, but not here. I would check consistency across all.

File: _topic_map.yml [hunk 17]
```diff
@@ -674,6 +674,9 @@ Topics:
 - Name: Kuryr SDN Administration
   File: kuryr
   Distros: openshift-origin
+- Name: Disabling Features using Feature Gates
+  File: disabling_features
+  Distros: openshift-origin,openshift-enterprise,openshift-dedicated
```
Comment (vikram-redhat) [inline _topic_map.yml]: @spurtell - would this feature be available in Dedicated docs at any point? 

@mburke5678

Comment (spurtell) [reply _topic_map.yml]: There are no plans to make this available in OpenShift Dedicated.

Comment (mburke5678) [reply _topic_map.yml]: Removed dedicated from topic map.
https://github.com/openshift/openshift-docs/pull/9856

**Comment by vikram-redhat** [comment]
@mburke5678 - has this been picked to 3.10?

**Comment by mburke5678** [comment]
@vikram-redhat Can you help? Why can't I cherrypick this or https://github.com/openshift/openshift-docs/pull/9856

**Comment by mburke5678** [comment]
>Assuming you'll disable builds, you'll see the builds tab in the web console.

@soltysh I have an issue asking about disabling builds using features gates based on the docs. I added [your example of disabling builds](https://github.com/openshift/openshift-docs/pull/8654#issuecomment-387072054). Is it possible to disable builds using a feature gate?

**Comment by soltysh** [comment]
From what I recall it should be possible.

**Pull request merged**